.github/workflows/zephyr.yml

```diff
@@ -7,6 +7,13 @@ name: Zephyr
 # yamllint disable-line rule:truthy
 on: [push, pull_request, workflow_dispatch, workflow_call]
 
+# Specifies group name that stops previous wokrflows if the name matches
+concurrency:
+  # eg. "Zephyr-pull_request-my_fork_branch_to_merge"
+  # eg. "Zephyr-push-refs/heads/my_branch_merging"
+  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.ref }}
```

Comment:
This one is actually stolen from Zephyr project and is a very smart one ;-)
This part:
`${{ github.head_ref || github.ref }}`
is a "[fallback value](https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-a-fallback-value)". Let me quote github.ref value docs:

> The fully-formed ref of the branch or tag that triggered the workflow run. For workflows triggered by push, this is the branch or tag ref that was pushed. For workflows triggered by pull_request, this is the pull request merge branch. For workflows triggered by release, this is the release tag created. For other triggers, this is the branch or tag ref that triggered the workflow run. This is only set if a branch or tag is available for the event type. The ref given is fully-formed, meaning that for branches the format is refs/heads/<branch_name>, for pull requests it is refs/pull/<pr_number>/merge, and for tags it is refs/tags/<tag_name>. For example, refs/heads/feature-branch-1.

So, github.ref would be valid for both "push" and "pull_request" events.
Right now we have a duplication in the events triggering - every PR starts a Zephyr workflow for both "push" and "pull_request" events what is a colossal waist of resources.
So thanks to using ${{ github.head_ref || github.ref }} we avoid duplication of those events in the PRs.
What is happening in detail:
1. You push to fork branch X on created PR to some other branch
2. Both event triggers for Zephyr workflows are valid - "on: push: and "on: pull_request", so they start two copies of the workflow (current implementation).
3. This concurrency group sees, that "on: push" event trigger has valid "github.head_ref" property since the push is done to the branch that has pull request assigned.
4. Thanks to this, workflow triggered by "on: push" has the very same group name that workflow created by "on: pull_request" event.
5. Workflows with group names matching are not being run accourding to concurrency keyword definition.
So, in the pull request only workflows triggered by "on: pull_request" are being launched.

What happens in other scenarios (non Pull Request):
1. When you merge a pull request or run daily tests triggered by scheduler using "on: workflow_call" or invoke Github Action manually using "on: workflow_dispatch", concurrency group name does not have a valid ${{ github.head_ref }} property.
2. Fallback value is used ${{ github.ref }} to establish group name.
3. If you retrigger the run - rerun every action on the very same commit it will cancel old run and start a new one as the group names match.

> "This is an interesting optimization but AFAIK it's only going to save "free" resources."

This saves not only "free" resources, it saves a really lot of resources.
I often wait for CI results because all Github runners are busy and not picking up workflows from my PRs due to high traffic. Now thanks to deduplication of the "pull_request" and "push" events we have more free runners for the project.
Secondly, I often wait for CI to pick up my workflows on PRs because I push frequently and impatiently.
And when I push to fix a single typo I need to wait for all my previous workflows to complete to start new ones.
This optimization  would stop all my previous running workflows and start new ones without waiting.
This is really useful feature (which I would not be smart enough to implement - as previously said stolen from Zephyr project).

Reply:
This should be probably copy pasted to all other workflows in our project in some other PR.